File: testdata/fixtures/java/CWE-79.java
package fixture;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

class Cwe79 {

    void viaWriter(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String x = request.getParameter("q");
        // ruleid: java-reflected-xss
        response.getWriter().write(x);
    }

    void viaPrint(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String name = request.getHeader("X-Name");
        // ruleid: java-reflected-xss
        response.getWriter().print(name);
    }

    void viaPrintln(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String data = request.getQueryString();
        // ruleid: java-reflected-xss
        response.getWriter().println(data);
    }

    void viaWriterConcat(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String val = request.getParameter("val");
        String out = "<div>" + val + "</div>";
        // ruleid: java-reflected-xss
        response.getWriter().write(out);
    }

    // ok: java-reflected-xss
    void safeConstant(HttpServletRequest request, HttpServletResponse response) throws Exception {
        response.getWriter().write("hello world");
    }

    // ok: java-reflected-xss
    void safeEncoded(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String x = request.getParameter("q");
        String safe = org.apache.commons.text.StringEscapeUtils.escapeHtml4("constant");
        response.getWriter().write(safe);
    }

    void safeFromConfig(String configured, HttpServletResponse response) throws Exception {
        // ok: java-reflected-xss
        response.getWriter().write(configured);
    }
}

File: testdata/fixtures/go/CWE-79.go
package fixture

import (
	"html/template"
	"net/http"
)

func viaTemplateHTML(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query().Get("q")
	// ruleid: go-reflected-xss
	t := template.HTML(q)
	_ = t
}

func viaWrite(w http.ResponseWriter, r *http.Request) {
	name := r.FormValue("name")
	// ruleid: go-reflected-xss
	w.Write([]byte(name))
}

func viaTemplateHTMLHeader(w http.ResponseWriter, r *http.Request) {
	h := r.Header.Get("X-Name")
	// ruleid: go-reflected-xss
	out := template.HTML(h)
	_ = out
}

func viaWriteConcat(w http.ResponseWriter, r *http.Request) {
	val := r.URL.Query().Get("val")
	out := "<div>" + val + "</div>"
	// ruleid: go-reflected-xss
	w.Write([]byte(out))
}

// ok: go-reflected-xss
func safeConstant(w http.ResponseWriter) {
	t := template.HTML("hello")
	_ = t
	w.Write([]byte("hello"))
}

// ok: go-reflected-xss
func safeEscaped(w http.ResponseWriter) {
	safe := template.HTMLEscapeString("constant")
	w.Write([]byte(safe))
}

func safeFromConfig(w http.ResponseWriter, data string) {
	// ok: go-reflected-xss
	t := template.HTML(data)
	_ = t
}
